tests/test_salcommander_minimal_controllers.py

```diff
@@ -0,0 +1,145 @@
+# including starting a CSC or loading a script (seconds)
+STD_TIMEOUT = 60
+
+INITIAL_LOG_LEVEL = 20
+
+index_gen = salobj.index_generator()
+
+languages = ["cpp", "java"]
+
+
+class SalCommanderToMinimalControllerTestCase(unittest.TestCase):
+    """Test a C++ or Java Commander communicating to
+    various minimal controllers."""
+
+    def setUp(self) -> None:
+        salobj.set_random_lsst_dds_partition_prefix()
+        self.index = next(index_gen)
+
+    def getLanguageVariables(self, language) -> None:
+        """Get the language specific test values;
+        either cpp (C++) or java (Java).
+
+        Parameters
+        ----------
+        language : `str`
+            Options are "java" or "cpp" and are case insensitive.
+
+        """
+        if language.lower() == "cpp":
+            commander = "minimal_cpp_commander.sh"
+            language_upper = "C++"
+            initial_log = "initial_log_level="
+        elif language.lower() == "java":
+            commander = "minimal_java_commander.sh"
+            language_upper = "Java"
+            initial_log = "initial logLevel.level "
+        else:
+            raise ValueError(f"Argument {language} must be one of 'cpp' or 'java'")
+        return commander, language_upper, initial_log
+
+    def test_cpp_controller(self) -> None:
+        for language in languages:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_cpp_controller.sh")
+
+    def test_java_controller(self) -> None:
+        for language in languages:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_java_controller.sh")
+
+    def test_salobj_controller(self) -> None:
+        for language in languages:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_salobj_controller.py")
+
+    def check_minimal_controller(self, language: str, exec_name: str) -> None:
+        # Start the Commander before the Controller subprocess
+        # to be sure the Commander sees telemetry from the Controller
+        # (telemetry is volatile, so has no historical data).
+        commander, language_upper, initial_log = self.getLanguageVariables(language)
+        commander_path = pathlib.Path(__file__).home() / "repos/ts_sal/bin" / commander
+        print(f"{language_upper} Commander: start {commander_path} with index={self.index} in a subprocess")
+        assert commander_path.is_file()
+        commander_process = subprocess.Popen(
+            [
+                commander_path,
+                str(self.index),
+                str(INITIAL_LOG_LEVEL),
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+        )
+        controller_path = (
+            pathlib.Path(__file__).parent.absolute() / "controllers" / exec_name
+        )
+        assert controller_path.is_file()
```

Comment:
I suggest testing this path before you launch the commander process (i.e. at about line 92). That avoids needlessly launching a process that you can't use.

Reply:
Fixed